Update config documentation (#452)
yiisoft/docs: src/guide/concept/configuration.md

the [default project template](https://github.com/yiisoft/app).

Yii3 configs are part of the application. You can change many aspects of how the application works by editing
configuration under `config/` directory.

## Config plugin

> The application template uses [yiisoft/config](https://github.com/yiisoft/config). It has two parts:

Reviewer: Is the intro about "two parts" needed at all? Both are described in detail just a bit later.

Author: I think yes, it looks more clearer

> 1. A Composer plugin that scans installed packages for configuration metadata and builds a merge plan.

Reviewer: ```suggestion
1. A Composer plugin that scans installed packages for configurations and builds a merge plan that declares the order of merging found configurations into final configuration to be passed to [DI container](di-container.md).
```

Reviewer: 1. It's not metadata, it is configs as well.
2. The purpose of merge plan was not clear.

Author: Updated

> 2. A config loader that reads a config group at runtime according to that plan.

Reviewer: What's group? Which group does it read?

Author: Updated

> Packages expose default config groups via the `config-plugin` key in the `extra` section of `composer.json`:

Reviewer: What's default config group?

Author: https://github.com/yiisoft/config#config-groups
Updated

```json
"extra": {
    "config-plugin-options": {
        "source-directory": "config",
    },

>     "config-plugin": {

Author: This is an example with a section of json schema.

        "params": "params.php",
        "web": "web.php"

>     }

Author: This is an example of a default configuration declared in a dependency package.

}
```

After Composer updates autoload files, such as during `dump-autoload`, `install`, `require`, `update`, or `remove`,

> the plugin scans installed packages and writes a merge plan to `config/.merge-plan.php`.

Reviewer: ```suggestion
the plugin scans installed packages for configurations and writes a merge plan to `config/.merge-plan.php`.
```

> At runtime, `Yiisoft\Config\Config` loads a config group according to that plan. Configs are read in three layers:

Reviewer: What's config group?

        'di' => 'common/di/*.php',
        'di-web' => [
            '$di',
            'web/di/*.php',
        ],
        'di-console' => '$di',
        'di-delegates' => [],
        'di-delegates-console' => '$di-delegates',
        'di-delegates-web' => '$di-delegates',
        'di-providers' => [],
        'di-providers-web' => '$di-providers',
        'di-providers-console' => '$di-providers',
        'events' => [],
        'events-web' => '$events',
        'events-console' => '$events',
        'routes' => 'common/routes.php',
    ],
    'config-plugin-environments' => [
        Environment::DEV => [
            'params' => [
                'environments/dev/params.php',
            ],
        ],
        Environment::TEST => [
            'params' => [
                'environments/test/params.php',
            ],
        ],
        Environment::PROD => [
            'params' => [
                'environments/prod/params.php',
            ],
        ],
    ],
];
```

Config group names are mapping keys. Their values are file paths relative to the `source-directory` option, which
defaults to the directory containing `composer.json`.

- A string value loads one file.

> - An array loads multiple files in the given order.

Reviewer: ```suggestion
- An array loads and merges multiple files in the given order.
```

- `?` marks an optional file.
- `*` marks a wildcard path.

> - `$group` references another config group.

Reviewer: ```suggestion
- `$group` references another config group to be merged.
```

For convenience, there is a naming convention about parameters:

1. Group parameters package name such as `yiisoft/cache-file`.
2. In case parameters are for the application itself, as in `application`, skip package prefix.
3. In case there are many services in the package, such as `file-target` and `file-rotator` in `yiisoft/log-target-file`
   package, group parameters by service name.
4. Use `enabled` as parameter name to be able to disable or enable a service, such as `yiisoft/yii-debug`.

### Package configs

> Config plugin does not copy package configs automatically. It loads vendor package configs directly from `vendor/`

Reviewer: ```suggestion
Config plugin does not copy vendor package configs automatically. It loads them directly from `vendor/`
```
